guardian/mobile-save-for-later
# Add @guardian/cdk to repository and deploy stage-specific templates

## What does this change?

We are migrating the `mobile-save-for-later` service from CloudFormation to `@guardian/cdk`. The migration will involve 5 PRs:

1. **Add `@guardian/cdk` to repository; update CI & deployment wiring accordingly [this PR]**
2. Provision a new version of the infrastructure using `@guardian/cdk` (https://github.com/guardian/mobile-save-for-later/pull/65)
3. Move DNS configuration from CloudFormation into `@guardian/cdk` (but continue routing requests to old infrastructure) (https://github.com/guardian/mobile-save-for-later/pull/66)
4. Update DNS configuration (start routing requests to new infrastructure) (https://github.com/guardian/mobile-save-for-later/pull/67)
5. Remove old infrastructure

In this PR, we 'wrap' the existing CloudFormation with `@guardian/cdk`, following the process described [here](https://github.com/guardian/cdk/blob/main/docs/migration-guide.md#initial-setup). 

Note that all changes in https://github.com/guardian/mobile-save-for-later/pull/64/commits/6179bb06ea21188929edad28a054cf3f33147aed are auto-generated by `@guardian/cdk` project creation tooling; I would not spend much time reviewing this in detail!

There are, however, a couple of changes that require a little more explanation: 

1. Original CloudFormation has been converted from `AWS::Serverless-2016-10-31` to 'standard' CloudFormation (https://github.com/guardian/mobile-save-for-later/pull/64/commits/1389750727148ad21fbe6205f45d27b757e4c9e2). 
    * This is to work around an issue described in this comment: https://github.com/aws/aws-cdk/issues/16654#issuecomment-929683266. In order to do this I downloaded the processed template via the AWS UI. 
2. We now deploy two separate templates (i.e. there is one file for `CODE` and a different file for `PROD`), instead of a single template for both stages (https://github.com/guardian/mobile-save-for-later/pull/64/commits/ad96d92d7c997a404d86f9f841e62ddd6470554a). 
    * This is the way that `@guardian/cdk` has chosen to deal with multiple stages. For motivation see: https://github.com/guardian/cdk/blob/main/docs/architecture-decision-records/006-stage-awareness.md. As it stands the contents of these templates are identical, but that will change in the second PR of this migration process. 

## How to test

I've deployed this to `CODE` and used a debug version of the app to confirm that saving articles/viewing saved articles still works as expected. (I checked the AWS Lambda logs and DynamoDB to confirm that server side code was being exercised as expected).

## How can we measure success?

We have started the migration process to `@guardian/cdk`.

## Have we considered potential risks?

This PR should be pretty low risk; we are changing CI and CD wiring, but the actual infrastructure serving requests will not change substantially (i.e. we just change the metadata by adding some new tags).
